### PR TITLE
refactor(grpc): add a flag for initial metadata to GetRequestMetadataFromContext

### PR DESCRIPTION
**Pull request by alevenberg**
https://github.com/googleapis/google-cloud-cpp/issues/13724

Add an is_initial_metadata_ready param to GetRequestMetadataFromContext. Set the default parameter as true since that is the current behavior.

This flag will be set by the IsClientOrign(Status s) call.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14104)

### CODE REVIEW

File: google/cloud/internal/async_read_write_stream_impl.h [hunk 1]
```diff
@@ -163,7 +163,8 @@ class AsyncStreamingReadWriteRpcImpl
   }
 
   RpcMetadata GetRequestMetadata() const override {
-    return GetRequestMetadataFromContext(*context_);
+    return GetRequestMetadataFromContext(*context_,
+                                         /*is_initial_metadata_ready=*/false);
```
Comment (dbolduc) [inline google/cloud/internal/async_read_write_stream_impl.h:167]: s/false/true/

File: google/cloud/internal/async_streaming_read_rpc_impl.h [hunk 2]
```diff
@@ -121,7 +121,8 @@ class AsyncStreamingReadRpcImpl : public AsyncStreamingReadRpc<Response> {
   }
 
   RpcMetadata GetRequestMetadata() const override {
-    return GetRequestMetadataFromContext(*context_);
+    return GetRequestMetadataFromContext(*context_,
+                                         /*is_initial_metadata_ready=*/false);
```
Comment (dbolduc) [inline google/cloud/internal/async_streaming_read_rpc_impl.h:125]: same.

File: google/cloud/testing_util/validate_metadata.cc [hunk 3]
```diff
@@ -248,9 +248,10 @@ RoutingHeaders ExtractRoutingHeaders(
  *   cannot reuse @p context for other RPCs or other calls to this function.
  */
 void SetServerMetadata(grpc::ClientContext& context,
-                       RpcMetadata const& server_metadata) {
+                       RpcMetadata const& server_metadata,
+                       bool is_initial_metadata_ready) {
```
Comment (dbolduc) [inline google/cloud/testing_util/validate_metadata.cc:252]: We should not make these changes to `testing_util`.

File: google/cloud/testing_util/validate_metadata.cc [hunk 4]
```diff
@@ -333,11 +334,14 @@ absl::optional<std::string> ValidateMetadataFixture::GetAuthority(
...
 
 void ValidateMetadataFixture::SetServerMetadata(
-    grpc::ClientContext& client_context, RpcMetadata const& server_metadata) {
+    grpc::ClientContext& client_context, RpcMetadata const& server_metadata,
+    bool is_initial_metadata_ready) {
   // Create a server context with the given server metadata.
   grpc::GenericServerContext server_context;
-  for (auto const& kv : server_metadata.headers) {
-    server_context.AddInitialMetadata(kv.first, kv.second);
+  if (is_initial_metadata_ready) {
```
Comment (dbolduc) [inline google/cloud/testing_util/validate_metadata.cc:341]: This thing is still exchanging the metadata. `grpc::ClientContext::initial_metadata_received_` will still be `true`.

All you are saying is to skip adding the metadata, which the caller could decide to do by leaving `server_metadata.headers` empty.

File: google/cloud/internal/grpc_request_metadata_test.cc [hunk 5]
```diff
@@ -46,6 +47,27 @@ TEST(GrpcRequestMetadata, GetRequestMetadataFromContext) {
...
 }
 
+TEST(GrpcRequestMetadata,
+     GetRequestMetadataFromContextWhenInitialMetadataIsNotReady) {
+  auto const server_metadata =
+      RpcMetadata{{{"header1", "value1"}, {"header2", "value2"}},
+                  {{"trailer1", "value3"}, {"trailer2", "value4"}}};
+  grpc::ClientContext context;
+  testing_util::SetServerMetadata(context, server_metadata,
+                                  /*is_initial_metadata_ready*/ false);
```
Comment (dbolduc) [inline google/cloud/internal/grpc_request_metadata_test.cc:57]: If you want to test that this case does not crash and burn, you should remove this call.

File: google/cloud/internal/grpc_request_metadata_test.cc [hunk 6]
```diff
@@ -46,6 +47,27 @@ TEST(GrpcRequestMetadata, GetRequestMetadataFromContext) {
...
+
+  auto md = GetRequestMetadataFromContext(context,
+                                          /*is_initial_metadata_ready*/ false);
```
Comment (dbolduc) [inline google/cloud/internal/grpc_request_metadata_test.cc:60]: nit: adding an `=` would be consistent. `/*is_initial_metadata_ready=*/`

File: google/cloud/internal/grpc_request_metadata_test.cc [hunk 7]
```diff
@@ -46,6 +47,27 @@ TEST(GrpcRequestMetadata, GetRequestMetadataFromContext) {
...
+  EXPECT_THAT(md.headers,
+              UnorderedElementsAre(
+                  // This function also returns the peer and compression
+                  // algorithm as synthetic headers.
+                  Pair(":grpc-context-peer", _),
+                  Pair(":grpc-context-compression-algorithm", "identity")));
+  EXPECT_THAT(md.trailers, UnorderedElementsAre(Pair("trailer1", "value3"),
+                                                Pair("trailer2", "value4")));
```
Comment (dbolduc) [inline google/cloud/internal/grpc_request_metadata_test.cc:68]: There will be no trailers, but that is probably what happens in the real world when we don't get any headers to begin with.

**Review by dbolduc — CHANGES_REQUESTED** [review]
Also @scotthart would say: consider using an `enum class` instead of a bool to improve callsite readability.

go/totw/94

Maybe like:

```cc
enum class GrpcMetadataView { kWithHeaders, kWithoutHeaders };
```

(optional)

File: google/cloud/internal/grpc_request_metadata_test.cc (hunk 5 above)
Comment (alevenberg) [reply google/cloud/internal/grpc_request_metadata_test.cc]: Done

File: google/cloud/internal/grpc_request_metadata_test.cc (hunk 7 above)
Comment (alevenberg) [reply google/cloud/internal/grpc_request_metadata_test.cc]: Ack.

File: google/cloud/testing_util/validate_metadata.cc (hunk 3 above)
Comment (alevenberg) [reply google/cloud/testing_util/validate_metadata.cc]: Done

File: google/cloud/testing_util/validate_metadata.cc (hunk 4 above)
Comment (alevenberg) [reply google/cloud/testing_util/validate_metadata.cc]: I removed these changes, so I think this is no longer relevant, but lmk if I am wrong

File: google/cloud/internal/status_utils.h [hunk 8]
```diff
@@ -23,7 +23,9 @@ namespace cloud {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace internal {
 
-bool IsClientOrigin(Status const& status);
+enum class ErrorOrigin { kUnknown, kClient };
+
+ErrorOrigin IsClientOrigin(Status const& status);
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/internal/status_utils.h]: Good catch. Changed.

File: google/cloud/testing_util/validate_metadata.h [hunk 9]
```diff
@@ -88,7 +88,8 @@ class ValidateMetadataFixture {
    *   cannot reuse @p context for other RPCs or other calls to this function.
    */
   void SetServerMetadata(grpc::ClientContext& client_context,
-                         RpcMetadata const& server_metadata = {});
+                         RpcMetadata const& server_metadata = {},
+                         bool error_origin = true);
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/testing_util/validate_metadata.h]: Ugh. missed it. Should be removed. We aren't changing the test infra anymore.

File: google/cloud/internal/async_read_write_stream_impl.h [hunk 10]
```diff
@@ -163,7 +164,9 @@ class AsyncStreamingReadWriteRpcImpl
   }
 
   RpcMetadata GetRequestMetadata() const override {
-    return GetRequestMetadataFromContext(*context_);
+    return GetRequestMetadataFromContext(
+        *context_,
+        /*error_origin=*/ErrorOrigin::kUnknown);
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/internal/async_read_write_stream_impl.h]: Done

**Comment by alevenberg** [comment]
I accidentally created this branch on the main repo. Going to close and re-open to fix.

**Pull request closed without merging**